File: examples/observability-cascade/observability-cascade-node-csharp/api/PublicApi.cs
using Holons.Observability;
using Relay.V1;

namespace CascadeNode.Csharp.Api;

public static class PublicApi
{
    public static TickResponse Tick(TickRequest request)
    {
        var obs = ObservabilityRegistry.Current();
        var slug = ResponderSlug(obs);
        var uid = obs.Config.InstanceUid ?? string.Empty;
        obs.Logger("tick").Info(
            "tick received",
            new Dictionary<string, object?>
            {
                ["sender"] = request.Sender,
                ["note"] = request.Note,
                ["responder_slug"] = slug,
                ["responder_uid"] = uid,
            });
        obs.Counter(
            "cascade_ticks_total",
            "Ticks received by this cascade node.",
            new Dictionary<string, string> { ["responder_uid"] = uid })?.Inc();
        return new TickResponse
        {
            ResponderSlug = slug,
            ResponderInstanceUid = uid,
        };
    }

    private static string ResponderSlug(Observability obs)
    {
        var configured = obs.Config.Slug?.Trim() ?? string.Empty;
        return string.IsNullOrEmpty(configured) ? "observability-cascade-node-csharp" : configured;
    }
}

File: examples/observability-cascade/observability-cascade-swift-node/Sources/CascadeNodeSwift/PublicAPI.swift
import Foundation
import Holons

public enum PublicAPI {
    public static func tick(_ request: Relay_V1_TickRequest) -> Relay_V1_TickResponse {
        let obs = current()
        let slug = responderSlug(obs)
        let uid = obs.cfg.instanceUid

        obs.logger("tick").info(
            "tick received",
            [
                "sender": request.sender,
                "note": request.note,
                "responder_slug": slug,
                "responder_uid": uid,
            ]
        )
        obs.counter(
            "cascade_ticks_total",
            help: "Ticks received by this cascade node.",
            labels: ["responder_uid": uid]
        )?.inc()

        var response = Relay_V1_TickResponse()
        response.responderSlug = slug
        response.responderInstanceUid = uid
        return response
    }

    private static func responderSlug(_ obs: Observability) -> String {
        let slug = obs.cfg.slug.trimmingCharacters(in: .whitespacesAndNewlines)
        return slug.isEmpty ? "observability-cascade-swift-node" : slug
    }
}
